tsdb/tsi1/report.go

```diff
@@ -0,0 +1,375 @@
+package tsi1
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"math"
+	"os"
+	"path/filepath"
+	"sort"
+	"text/tabwriter"
+	"time"
+
+	"github.com/influxdata/influxdb"
+	"github.com/influxdata/influxdb/tsdb"
+	"go.uber.org/zap"
+)
+
+const (
+	// Number of series IDs to stored in slice before we convert to a roaring
+	// bitmap. Roaring bitmaps have a non-trivial initial cost to construct.
+	useBitmapN = 25
+)
+
+// ReportCommand represents the program execution for "influxd inspect report-tsi".
+type ReportCommand struct {
+	// Standard input/output, overridden for testing.
+	Stderr io.Writer
+	Stdout io.Writer
+	Logger *zap.Logger
+
+	// Filters
+	DataPath        string
+	OrgID, BucketID *influxdb.ID
+
+	byOrgBucket         map[influxdb.ID]map[influxdb.ID]*cardinality
+	byBucketMeasurement map[influxdb.ID]map[string]*cardinality
+	orgToBucket         map[influxdb.ID][]influxdb.ID
+
+	SeriesDirPath string // optional. Defaults to dbPath/_series
+	sfile         *tsdb.SeriesFile
+	indexFile     *Index
+
+	TopN          int
+	ByMeasurement bool
+	byTagKey      bool
+
+	start time.Time
+}
+
+// NewReportCommand returns a new instance of ReportCommand with default setting applied.
+func NewReportCommand() *ReportCommand {
+	return &ReportCommand{
+		Logger:              zap.NewNop(),
+		byOrgBucket:         make(map[influxdb.ID]map[influxdb.ID]*cardinality),
+		byBucketMeasurement: make(map[influxdb.ID]map[string]*cardinality),
+		orgToBucket:         make(map[influxdb.ID][]influxdb.ID),
+		TopN:                0,
+		byTagKey:            false,
+	}
+}
+
+// ReportTSISummary is returned by a report-tsi Run() command and is used to access cardinality information
+type Summary struct {
+	TotalCardinality             int64
+	OrgCardinality               map[influxdb.ID]int64
+	BucketByOrgCardinality       map[influxdb.ID]map[influxdb.ID]int64
+	BucketMeasurementCardinality map[influxdb.ID]map[string]int64
+}
+
+func newSummary() *Summary {
+	return &Summary{
+		OrgCardinality:               make(map[influxdb.ID]int64),
+		BucketByOrgCardinality:       make(map[influxdb.ID]map[influxdb.ID]int64),
+		BucketMeasurementCardinality: make(map[influxdb.ID]map[string]int64),
+	}
+}
+
+// Run runs the report-tsi tool which can be used to find the cardinality
+// any org or bucket. Run returns a *ReportTSISummary, which contains maps for finding
+// the cardinality of a bucket or org based on its influxdb.ID
+func (report *ReportCommand) Run(print bool) (*Summary, error) {
+	report.start = time.Now()
+
+	report.Stdout = os.Stdout
+
+	if report.SeriesDirPath == "" {
```

Comment:
I think we can just have the `_series` path be explicit rather than inferring it.